### PR TITLE
Add resources and datasources named vkcs_kubernetes_cluster_v2, vkcs_kubernetes_node_group_v2

### PR DESCRIPTION
**Pull request by cantylv**

### CODE REVIEW

File: examples/kubernetes/cluster_v2/regional/datasource.tf [hunk 1]
```diff
@@ -0,0 +1,3 @@
+data "vkcs_kubernetes_cluster_v2" "cluster" {
+  id = "39UdIv4W0EegBs2EYVeGdas38do"
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/regional/datasource.tf]: You can't hardcode identifier here. These examples are being tested. We usually use reference to resource which is placed in the same directory.

File: examples/kubernetes/cluster_v2/regional/flavor.tf [hunk 2]
```diff
@@ -0,0 +1,3 @@
+data "vkcs_compute_flavor" "master" {
+  name = "STD2-8-8"
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/regional/flavor.tf]: We usually use symbolic links referencing common infrastructure resources. Check for "base-flavor.tf".

File: examples/kubernetes/cluster_v2/regional/main.tf [hunk 3]
```diff
@@ -0,0 +1,31 @@
+resource "vkcs_kubernetes_cluster_v2" "cluster" {
+  name        = "base-infra"
+  description = "An example of a Kubernetes cluster v2 created via Terraform"
+  version     = "v1.31.4"
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/regional/main.tf]: It would be better to use the latest supported version here.

File: examples/kubernetes/cluster_v2/regional/network.tf [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# Create networks
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/regional/network.tf]: Same as with the flavor above, check for "base-networking.tf".

File: examples/kubernetes/cluster_v2/regional/outputs.tf [hunk 5]
```diff
@@ -0,0 +1,144 @@
+output "cluster_id" {
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/regional/outputs.tf]: We usually use outputs in examples only for specific user scenarios. I don't think we need an output for each cluster's attribute.

File: examples/kubernetes/cluster_v2/standard/datasource.tf [hunk 6]
```diff
@@ -0,0 +1,3 @@
+data "vkcs_kubernetes_cluster_v2" "cluster" {
+  id = "39UdIv4W0EegBs2EYVeGdas38do"
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/standard/datasource.tf]: Same as above.

File: examples/kubernetes/cluster_v2/standard/flavor.tf [hunk 7]
```diff
@@ -0,0 +1,3 @@
+data "vkcs_compute_flavor" "master" {
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/standard/flavor.tf]: Same as above.

File: examples/kubernetes/cluster_v2/regional/main.tf [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  enable_public_ip = true
+
+  labels = {
+    managed-by = "terraform"
+  }
+
+  insecure_registries = ["registry.example.com:5000"]
+
+  loadbalancer_subnet_id     = vkcs_networking_subnet.k8s-subnet.id
+  loadbalancer_allowed_cidrs = ["10.0.0.0/8", "192.168.0.0/16"]
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/regional/main.tf]: What's the use case?

File: examples/kubernetes/cluster_v2/regional/main.tf [hunk 9]
```diff
@@ -0,0 +1,31 @@
+resource "vkcs_kubernetes_cluster_v2" "cluster" {
+  name        = "base-infra"
+  description = "An example of a Kubernetes cluster v2 created via Terraform"
+  version     = "v1.31.4"
+
+  cluster_type = "regional"
+  master_count = 5
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/regional/main.tf]: Isn't it too much? It's better to keep examples minimal, isn't 3 enough?

File: examples/kubernetes/node_group_v2/cluster.tf [hunk 10]
```diff
@@ -0,0 +1,31 @@
+resource "vkcs_kubernetes_cluster_v2" "cluster" {
```
Comment (paaanic) [inline examples/kubernetes/node_group_v2/cluster.tf]: It's better to use a symbolic link here as well.

File: examples/kubernetes/node_group_v2/main.tf [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  availability_zone = "GZ1"
+
+  scale_type = "fixed_scale"
+  node_count = 1
+
+  labels = {
+    app = "web-service"
+  }
+
+  dynamic "taints" {
```
Comment (paaanic) [inline examples/kubernetes/node_group_v2/main.tf]: What's the reason for this to be dynamic? Can't we just specify two taints? And, btw, we don't use blocks in framework resources, you should use list attribute instead.

File: examples/kubernetes/cluster_v2/standard/main.tf [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+  cluster_type = "standard"
+  master_count = 1
+
+  master_flavor = data.vkcs_compute_flavor.master.id
+
+  network_id = vkcs_networking_network.k8s-network.id
+  subnet_id  = vkcs_networking_subnet.k8s-subnet.id
+
+  availability_zones = ["GZ1"]
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/standard/main.tf]: Probably not the best choice for AZ.

File: templates/kubernetes/data-sources/vkcs_kubernetes_cluster_v2.md.tmpl [hunk 13]
```diff
@@ -0,0 +1,16 @@
+---
+subcategory: "{{.SubCategory}}"
+layout: "vkcs"
+page_title: "vkcs: {{.Name}}"
+description: |-
+  Get information on cluster v2.
```
Comment (paaanic) [inline templates/kubernetes/data-sources/vkcs_kubernetes_cluster_v2.md.tmpl]: It would be better to use the naming corresponding to how it's named in the web UI. I suppose it's something like "new generation".

File: vkcs/internal/clients/client.go [hunk 14]
```diff
@@ -57,6 +57,10 @@ func newContainerInfraV1(client *gophercloud.ProviderClient, opts clientOpts) (*
 	return initClientOptsNew(client, opts, "container-infra")
 }
 
+func newContainerInfraV2(client *gophercloud.ProviderClient, opts clientOpts) (*gophercloud.ServiceClient, error) {
```
Comment (paaanic) [inline vkcs/internal/clients/client.go]: We use names from Keystone catalog here.

File: vkcs/kubernetes/v2/datasource_kubernetes_cluster_v2/converters.go [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+	"context"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/vk-cs/terraform-provider-vkcs/vkcs/internal/services/kubernetes/containerinfra/v2/clusters"
+)
+
+// FlattenLabels конвертирует map[string]string в types.Map
```
Comment (paaanic) [inline vkcs/kubernetes/v2/datasource_kubernetes_cluster_v2/converters.go]: We don't use Russian language in this project.

File: vkcs/kubernetes/datasource_kubernetes_cluster_v2/model_converters.go [hunk 16]
```diff
@@ -0,0 +1,219 @@
+package datasource_kubernetes_cluster_v2
```
Comment (paaanic) [inline vkcs/kubernetes/datasource_kubernetes_cluster_v2/model_converters.go:1]: You should place all of these files in the kubernetes directory, instead of creating a separate one for next-generation clusters. Acceptance testing mechanisms are based on the directories structure.

File: examples/kubernetes/cluster_v2/standard/main.tf [hunk 17]
```diff
@@ -0,0 +1,31 @@
+resource "vkcs_kubernetes_cluster_v2" "cluster" {
+  name        = "base-infra"
+  description = "An example of a Kubernetes cluster v2 created via Terraform"
+  version     = "v1.31.4"
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/standard/main.tf]: How user should get information on the list of available versions? We should add a data source for this.

File: vkcs/kubernetes/v2/resource_cluster_v2.go [hunk 18]
```diff
@@ -0,0 +1,586 @@
...
+			}
+			return cluster, cluster.Status, nil
+		},
+		Timeout:    operationCreateV2 * time.Minute,
+		Delay:      createClusterDelayV2 * time.Second,
+		MinTimeout: createClusterPollIntervalV2 * time.Second,
+	}
+}
+
+func (r *kubernetesClusterV2Resource) validateCluster(ctx context.Context, data *rkubeclusterv2.KubernetesClusterV2Model, diags *diag.Diagnostics) bool {
```
Comment (paaanic) [inline vkcs/kubernetes/v2/resource_cluster_v2.go]: All of these validations should be done in the configuration step by applying schema-based validators or with an use of `ValidateConfig` method (https://developer.hashicorp.com/terraform/plugin/framework/resources/validate-configuration).

File: vkcs/kubernetes/v2/resource_cluster_v2.go [hunk 19]
```diff
@@ -0,0 +1,586 @@
...
+		if dKubeVersion.HasError() {
+			diags.Append(dKubeVersion...)
+			return false
+		}
+	}
+
+	return true
+}
+
+func (r *kubernetesClusterV2Resource) validateAZs(region string, clusterAZs []string) (diags diag.Diagnostics) {
```
Comment (paaanic) [inline vkcs/kubernetes/v2/resource_cluster_v2.go]: I don't see a reason to validate it this way. API should provide an informative response on a bad client request, and this should be enough here.

File: vkcs/kubernetes/v2/resource_cluster_v2.go [hunk 20]
```diff
@@ -0,0 +1,586 @@
...
+		if _, ok := mk8sAZs[az]; !ok {
+			allowedAZs := r.extractKeysFromMap(mk8sAZs)
+			diags.AddError("Configuration of resource vkcs_kubernetes_cluster_v2 contains invalid availability zone", fmt.Sprintf("allowed %#v, got %#v", allowedAZs, clusterAZs))
+		}
+	}
+
+	return
+}
+
+func (r *kubernetesClusterV2Resource) validateKubeVersion(region string, kubeVersion string) (diags diag.Diagnostics) {
```
Comment (paaanic) [inline vkcs/kubernetes/v2/resource_cluster_v2.go]: Same as above.

File: vkcs/kubernetes/resource_cluster_v2.go [hunk 21]
```diff
@@ -0,0 +1,586 @@
...
+	// Validate cluster configuration
+	if !r.validateCluster(ctx, &plan, &resp.Diagnostics) {
+		return
+	}
+
+	// Validate simultaneous scaling and upgrade
+	versionChanged := !plan.Version.Equal(state.Version)
+	masterFlavorChanged := !plan.MasterFlavor.Equal(state.MasterFlavor)
+
+	if versionChanged && masterFlavorChanged {
```
Comment (paaanic) [inline vkcs/kubernetes/resource_cluster_v2.go]: It should be placed in validation during configuration step as well.

File: vkcs/kubernetes/resource_cluster_v2.go [hunk 22]
```diff
@@ -0,0 +1,586 @@
...
+	client, err := r.config.ContainerInfraV2Client(data.Region.ValueString())
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating Container Infra V2 API client for resource vkcs_kubernetes_cluster_v2", err.Error())
+		return
+	}
+
+	tflog.Trace(ctx, "Calling Container Infra V2 API to delete cluster", map[string]interface{}{"cluster_id": fmt.Sprintf("%#v", data.Id.ValueString())})
+
+	err = clusters.Delete(client, data.Id.ValueString())
+	if err != nil {
```
Comment (paaanic) [inline vkcs/kubernetes/resource_cluster_v2.go:440]: You should accept a not found error here.

File: vkcs/kubernetes/resource_cluster_v2.go [hunk 23]
```diff
@@ -0,0 +1,586 @@
...
+			clusterStatusV2Deleting,
+			clusterStatusV2Running,
+		},
+		Target: []string{
+			clusterStatusV2Deleted,
+		},
+		Refresh: func() (interface{}, string, error) {
+			cluster, err := clusters.Get(client, clusterID).Extract()
+			if err != nil {
+				if util.CheckDeletedStatus(err) {
```
Comment (paaanic) [inline vkcs/kubernetes/resource_cluster_v2.go]: I don's see a reason to add another utility. It's basically the same 404 check with a new name.

File: vkcs/kubernetes/resource_cluster_v2.go [hunk 24]
```diff
@@ -0,0 +1,586 @@
...
+					FlavorID: plan.MasterFlavor.ValueString(),
+				},
+			},
+			Replicas: int(plan.MasterCount.ValueInt64()),
+		}
+		scaleOpts.MasterSpec = masterSpec
+	}
+
+	// Only update if there are changes
+	if upgradeOpts.Version == "" && scaleOpts.MasterSpec.Engine.NovaEngine.FlavorID == "" {
```
Comment (paaanic) [inline vkcs/kubernetes/resource_cluster_v2.go]: It should be avoided when using a plan modifiers on other attributes instead of this "if" statement.

File: vkcs/kubernetes/resource_cluster_v2.go [hunk 25]
```diff
@@ -0,0 +1,586 @@
...
+	// Wait for update to complete
+	stateConf := r.getStateConfForClusterUpdate(client, plan.Id.ValueString())
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		resp.Diagnostics.AddError("Error waiting for Kubernetes cluster V2 update to complete", err.Error())
+		return
+	}
+
+	// Read the updated cluster
+	readClusterDiags := r.readCluster(ctx, client, &plan)
```
Comment (paaanic) [inline vkcs/kubernetes/resource_cluster_v2.go]: Using the plan to store the current API state is not the best way to do it. We usually read it once into an another variable after applying changes.

File: vkcs/kubernetes/v2/resource_node_group_v2.go [hunk 26]
```diff
@@ -0,0 +1,672 @@
...
+		Timeout:      30 * time.Minute,
+		Delay:        deleteNgDelayV2 * time.Second,
+		PollInterval: deleteNgPollIntervalV2 * time.Second,
+	}
+}
+
+func (r *kubernetesNodeGroupV2Resource) ImportState(ctx context.Context, req resource.ImportStateRequest, resp *resource.ImportStateResponse) {
+	parts := strings.Split(req.ID, "/")
+	if len(parts) != 2 {
+		resp.Diagnostics.AddError("Invalid import ID", "Import ID must be in the format 'cluster_id/node_group_id'")
```
Comment (paaanic) [inline vkcs/kubernetes/v2/resource_node_group_v2.go]: Why? You only use the identifier of a node group itself when reading it.

File: vkcs/kubernetes/spec.json [hunk 27]
```diff
@@ -0,0 +1,1488 @@
+{
+  "provider": {
+    "name": "vkcs"
+  },
+  "datasources": [
+    {
+      "name": "kubernetes_cluster_v2",
```
Comment (paaanic) [inline vkcs/kubernetes/spec.json:7]: I won't check the schema on per-attribute basis now, but I want to say that, as for now, we usually try to stick to the API schema here. And I see that they don't correspond well now. Also you should check that all attributes that are not marked as "requires replace" can be updated.

File: CHANGELOG.md [hunk 28]
```diff
@@ -6,6 +6,13 @@ description: |-
 ---
 
 # VKCS Provider's changelog
+#### v0.15.0
+- Add new client - managed-k8s
```
Comment (paaanic) [inline CHANGELOG.md]: It's not a change an user should be informed about.

**Review by paaanic — CHANGES_REQUESTED** [review]
Done the first round. I looked superficially, so there might be more changes requested upcoming. For now, you should work on schema and make it more API-like. We now tend to avoid make one much different from another, while it doesn't bother an user, because it usually leads to inability to implement something correctly in specific use-cases and brings in discord. Also, you should add acceptance tests for the added resources. Usually, to start, we add a basic test which covers the case with minimum attributes specified, a "full" test with all attributes specified, and a full update test in which all attributes are updated. Check for examples.

File: examples/kubernetes/cluster_v2/regional/datasource.tf (hunk 1 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/regional/datasource.tf]: okay, fixed

File: examples/kubernetes/cluster_v2/regional/flavor.tf (hunk 2 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/regional/flavor.tf]: fixed

File: examples/kubernetes/cluster_v2/regional/main.tf (hunk 3 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/regional/main.tf]: changed version to 1.34.2

File: examples/kubernetes/cluster_v2/regional/network.tf (hunk 4 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/regional/network.tf]: fixed

File: examples/kubernetes/cluster_v2/regional/outputs.tf (hunk 5 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/regional/outputs.tf]: agree, i've deleted outputs

File: examples/kubernetes/cluster_v2/standard/datasource.tf (hunk 6 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/standard/datasource.tf]: fixed

File: examples/kubernetes/cluster_v2/standard/flavor.tf (hunk 7 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/standard/flavor.tf]: fixed

File: examples/kubernetes/cluster_v2/regional/main.tf (hunk 9 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/regional/main.tf]: 3 is enough, i've changed to 3

File: examples/kubernetes/node_group_v2/cluster.tf (hunk 10 above)
Comment (cantylv) [reply examples/kubernetes/node_group_v2/cluster.tf]: fixed

File: examples/kubernetes/cluster_v2/regional/main.tf (hunk 8 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/regional/main.tf]: `loadbalancer_allowed_cidrs` parameter in a Kubernetes cluster is used to restrict access to the Kubernetes API through the load balancer. It determines which IP addresses or networks are allowed to connect to the cluster API.

File: examples/kubernetes/node_group_v2/main.tf (hunk 11 above)
Comment (cantylv) [reply examples/kubernetes/node_group_v2/main.tf]: fixed example

File: examples/kubernetes/cluster_v2/standard/main.tf (hunk 12 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/standard/main.tf]: a lot of examples use GZ1, maybe it's ok ?

File: templates/kubernetes/data-sources/vkcs_kubernetes_cluster_v2.md.tmpl (hunk 13 above)
Comment (cantylv) [reply templates/kubernetes/data-sources/vkcs_kubernetes_cluster_v2.md.tmpl]: fixed

File: vkcs/internal/clients/client.go (hunk 14 above)
Comment (cantylv) [reply vkcs/internal/clients/client.go]: fixed, now it's 'newManagedK8S'

File: vkcs/kubernetes/v2/datasource_kubernetes_cluster_v2/converters.go (hunk 15 above)
Comment (cantylv) [reply vkcs/kubernetes/v2/datasource_kubernetes_cluster_v2/converters.go]: ok, now there are only english comments

File: vkcs/kubernetes/datasource_kubernetes_cluster_v2/model_converters.go (hunk 16 above)
Comment (cantylv) [reply vkcs/kubernetes/datasource_kubernetes_cluster_v2/model_converters.go:1]: moved files to vkcs/kubernetes

now it's same as vkcs/cdn

File: vkcs/kubernetes/v2/resource_cluster_v2.go (hunk 18 above)
Comment (cantylv) [reply vkcs/kubernetes/v2/resource_cluster_v2.go]: i've chosen schema-based validation for attributes values

File: vkcs/kubernetes/v2/resource_cluster_v2.go (hunk 19 above)
Comment (cantylv) [reply vkcs/kubernetes/v2/resource_cluster_v2.go]: i agree
i've deleted these methods. i had idea to validate available zones and kube versions for cluster on stage 'terraform plan', but actually it's excessive.

File: vkcs/kubernetes/v2/resource_cluster_v2.go (hunk 20 above)
Comment (cantylv) [reply vkcs/kubernetes/v2/resource_cluster_v2.go]: i agree
i've deleted these methods. i had idea to validate available zones and kube versions for cluster on stage 'terraform plan', but actually it's excessive.

File: examples/kubernetes/cluster_v2/standard/main.tf (hunk 17 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/standard/main.tf]: add datasource vkcs_kubernetes_versions_v2

File: vkcs/kubernetes/resource_cluster_v2.go (hunk 21 above)
Comment (cantylv) [reply vkcs/kubernetes/resource_cluster_v2.go]: i've placed it in ModifyPlan, because there i can interact with plan and state. In validation block I don't have access to the state value.

File: vkcs/kubernetes/resource_cluster_v2.go (hunk 22 above)
Comment (cantylv) [reply vkcs/kubernetes/resource_cluster_v2.go:440]: fixed

File: vkcs/kubernetes/resource_cluster_v2.go (hunk 23 above)
Comment (cantylv) [reply vkcs/kubernetes/resource_cluster_v2.go]: agree, fixed

File: vkcs/kubernetes/resource_cluster_v2.go (hunk 24 above)
Comment (cantylv) [reply vkcs/kubernetes/resource_cluster_v2.go]: agree, fixed

File: vkcs/kubernetes/resource_cluster_v2.go (hunk 25 above)
Comment (cantylv) [reply vkcs/kubernetes/resource_cluster_v2.go]: agree, fixed

File: vkcs/kubernetes/v2/resource_node_group_v2.go (hunk 26 above)
Comment (cantylv) [reply vkcs/kubernetes/v2/resource_node_group_v2.go]: I copied it from documentation, my fault

fixed

File: vkcs/kubernetes/spec.json (hunk 27 above)
Comment (cantylv) [reply vkcs/kubernetes/spec.json:7]: Checked and fixed

File: CHANGELOG.md (hunk 28 above)
Comment (cantylv) [reply CHANGELOG.md]: fixed changelog

**Comment by cantylv** [comment]
> Done the first round. I looked superficially, so there might be more changes requested upcoming. For now, you should work on schema and make it more API-like. We now tend to avoid make one much different from another, while it doesn't bother an user, because it usually leads to inability to implement something correctly in specific use-cases and brings in discord. Also, you should add acceptance tests for the added resources. Usually, to start, we add a basic test which covers the case with minimum attributes specified, a "full" test with all attributes specified, and a full update test in which all attributes are updated. Check for examples.

results of acceptance tests:

➜  kubernetes git:(vkcsdev-vkcs-kubernetes-v2-spec) ✗ go test -v -run="TestAccKubernetesNodeGroupV2" -parallel 10 -timeout 60m
=== RUN   TestAccKubernetesNodeGroupV2DataSource_basic
=== PAUSE TestAccKubernetesNodeGroupV2DataSource_basic
=== RUN   TestAccKubernetesNodeGroupV2_basic
=== PAUSE TestAccKubernetesNodeGroupV2_basic
=== RUN   TestAccKubernetesNodeGroupV2_full
=== PAUSE TestAccKubernetesNodeGroupV2_full
=== RUN   TestAccKubernetesNodeGroupV2_update
=== PAUSE TestAccKubernetesNodeGroupV2_update
=== CONT  TestAccKubernetesNodeGroupV2DataSource_basic
=== CONT  TestAccKubernetesNodeGroupV2_full
=== CONT  TestAccKubernetesNodeGroupV2_basic
=== CONT  TestAccKubernetesNodeGroupV2_update
--- PASS: TestAccKubernetesNodeGroupV2DataSource_basic (1857.35s)
--- PASS: TestAccKubernetesNodeGroupV2_full (1898.93s)
--- PASS: TestAccKubernetesNodeGroupV2_basic (2001.78s)
--- PASS: TestAccKubernetesNodeGroupV2_update (2385.86s)
PASS
ok      github.com/vk-cs/terraform-provider-vkcs/vkcs/kubernetes        2386.639s



➜  kubernetes git:(vkcsdev-vkcs-kubernetes-v2-spec) ✗ go test -v -run="TestAccKubernetesClusterV2" -parallel 10 -timeout 60m
=== RUN   TestAccKubernetesClusterV2DataSource_basic
=== PAUSE TestAccKubernetesClusterV2DataSource_basic
=== RUN   TestAccKubernetesClusterV2_basic
=== PAUSE TestAccKubernetesClusterV2_basic
=== RUN   TestAccKubernetesClusterV2_full
=== PAUSE TestAccKubernetesClusterV2_full
=== RUN   TestAccKubernetesClusterV2_scale
=== PAUSE TestAccKubernetesClusterV2_scale
=== RUN   TestAccKubernetesClusterV2_upgrade
=== PAUSE TestAccKubernetesClusterV2_upgrade
=== CONT  TestAccKubernetesClusterV2DataSource_basic
=== CONT  TestAccKubernetesClusterV2_scale
=== CONT  TestAccKubernetesClusterV2_full
=== CONT  TestAccKubernetesClusterV2_basic
=== CONT  TestAccKubernetesClusterV2_upgrade
--- PASS: TestAccKubernetesClusterV2_full (1263.56s)
--- PASS: TestAccKubernetesClusterV2_basic (1273.22s)
--- PASS: TestAccKubernetesClusterV2DataSource_basic (1283.28s)
--- PASS: TestAccKubernetesClusterV2_scale (1622.56s)
--- PASS: TestAccKubernetesClusterV2_upgrade (2155.86s)
PASS
ok      github.com/vk-cs/terraform-provider-vkcs/vkcs/kubernetes        2157.227s

File: examples/kubernetes/cluster_v2/regional/main-datasource.tf [hunk 29]
```diff
@@ -0,0 +1,3 @@
+data "vkcs_kubernetes_cluster_v2" "k8s_cluster" {
+  id = vkcs_kubernetes_cluster.k8s_cluster.id
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/regional/main-datasource.tf:2]:
```suggestion
  id = vkcs_kubernetes_cluster_v2.k8s_cluster.id
```

File: examples/kubernetes/cluster_v2/regional/main.tf (hunk 8 above)
Comment (paaanic) [reply examples/kubernetes/cluster_v2/regional/main.tf]: We usually try to set up a minimal working example. I think it's a bit extra here. Does an user need to create subnets with these specified cidrs and make them routable with loadbalancer subnet to be allowed to work with Kube API?

File: CHANGELOG.md [hunk 30]
```diff
@@ -338,4 +346,4 @@ description: |-
 - Fix database instances import: databases & users
 
 #### v0.1.0
-- Initial release
+- Initial release
```
Comment (paaanic) [inline CHANGELOG.md:349]: What's changed here?

File: examples/kubernetes/cluster_v2/standard/main.tf [hunk 31]
```diff
@@ -0,0 +1,37 @@
+resource "vkcs_kubernetes_cluster_v2" "k8s_cluster" {
+  name        = "k8s-standard-cluster"
+  description = "An example of a standard Kubernetes cluster v2 created via Terraform"
+  version     = "v1.34.2"
+
+  cluster_type = "standard"
```
Comment (paaanic) [inline examples/kubernetes/cluster_v2/standard/main.tf]: Do we actually need this to distinguish? Also, it's better to stick with API naming convention here. It's named "deployment type" there as I see (zonal/multi_zonal).

File: vkcs/kubernetes/v2/resource_cluster_v2.go (hunk 19 above)
Comment (paaanic) [reply vkcs/kubernetes/v2/resource_cluster_v2.go]: Now it checks for string emptiness. I believe that can be achieved using base schema-based stringvalidator.

File: vkcs/kubernetes/resource_cluster_v2.go [hunk 32]
```diff
@@ -0,0 +1,474 @@
...
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *kubernetesClusterV2Resource) readCluster(client *gophercloud.ServiceClient, clusterID string) (cluster *clusters.Cluster, clusterKubeconfig *string, diags diag.Diagnostics) {
+	// Get cluster data
+	cluster, err := clusters.Get(client, clusterID).Extract()
+	if err != nil {
+		if errutil.IsNotFound(err) {
```
Comment (paaanic) [inline vkcs/kubernetes/resource_cluster_v2.go:294]: You cannot ignore it here. Instead you should remove the state of the resource: https://github.com/vk-cs/terraform-provider-vkcs/blob/master/vkcs/cdn/resource_origin_group.go#L127.

**Review by paaanic — CHANGES_REQUESTED** [review]

File: examples/kubernetes/cluster_v2/regional/main.tf (hunk 8 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/regional/main.tf]: No, the user doesn't need to create these specified CIDRs, etc. I thought we should introduce how to use the resource's arguments.

Comment (cantylv) [reply examples/kubernetes/cluster_v2/regional/main.tf]: I've deleted unnecessary arguments from node_group_v2 and cluster_v2

File: examples/kubernetes/cluster_v2/standard/main.tf (hunk 12 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/standard/main.tf]: Okay, I've changed it or "PA2", or "ME1", or "MS1"

File: examples/kubernetes/cluster_v2/regional/main-datasource.tf (hunk 29 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/regional/main-datasource.tf:2]: oops, fixed

File: vkcs/kubernetes/v2/resource_cluster_v2.go (hunk 19 above)
Comment (cantylv) [reply vkcs/kubernetes/v2/resource_cluster_v2.go]: fixed

File: CHANGELOG.md (hunk 30 above)
Comment (cantylv) [reply CHANGELOG.md]: The newline character is missing

Returned it

File: examples/kubernetes/cluster_v2/standard/main.tf (hunk 31 above)
Comment (cantylv) [reply examples/kubernetes/cluster_v2/standard/main.tf]: These are business requirements from the product manager, and the same thing was also done for the old‑generation cluster. I think that makes sense.

File: vkcs/kubernetes/resource_cluster_v2.go (hunk 32 above)
Comment (cantylv) [reply vkcs/kubernetes/resource_cluster_v2.go:339]: okay, fixed

File: vkcs/kubernetes/resource_node_group_v2.go [hunk 33]
```diff
@@ -0,0 +1,493 @@
...
+	if err != nil {
+		resp.Diagnostics.AddError("Error creating Kubernetes node group", err.Error())
+		return
+	}
+
+	// Set the ID immediately
+	data.Id = types.StringValue(nodeGroupID)
+
+	// Parse timeout for operation
+	createTimeout, err := time.ParseDuration(data.Timeouts.Create.ValueString())
```
Comment (paaanic) [inline vkcs/kubernetes/resource_node_group_v2.go]: You should use common type for timeouts attribute: github.com/hashicorp/terraform-plugin-framework-timeouts/resource/timeouts. Check for references.

File: vkcs/kubernetes/resource_cluster_v2_test.go [hunk 34]
```diff
@@ -0,0 +1,376 @@
...
+					testAccCheckKubernetesClusterV2Exists("vkcs_kubernetes_cluster_v2.scale"),
+					resource.TestCheckResourceAttr("vkcs_kubernetes_cluster_v2.scale", "name", clusterName),
+					resource.TestCheckResourceAttrPair("vkcs_kubernetes_cluster_v2.scale", "master_flavor", "data.vkcs_compute_flavor.base", "id"),
+				),
+			},
+		},
+	})
+}
+
+/*
```
Comment (paaanic) [inline vkcs/kubernetes/resource_cluster_v2_test.go]: It would be better to add a note why this is commented.

File: vkcs/provider/provider.go [hunk 35]
```diff
@@ -325,6 +331,10 @@ func (p *vkcsProvider) DataSources(_ context.Context) []func() datasource.DataSo
 		kubernetes.NewNodeGroupDataSource,
 		kubernetes.NewSecurityPolicyTemplatesDataSource,
 		kubernetes.NewSecurityPolicyTemplateDataSource,
+		kubernetes.NewKubernetesClusterV2DataSource,
+		kubernetes.NewKubernetesNodeGroupV2DataSource,
+		kubernetes.NewKubernetesVolumeTypesV2DataSource,
+		kubernetes.NewKubernetesClusterVersionV2DataSource,
```
Comment (paaanic) [inline vkcs/provider/provider.go]: nit: We usually try to maintain this list sorted.

File: vkcs/internal/framework/validators/uuid.go
Comment (paaanic) [inline vkcs/internal/framework/validators/uuid.go:1]: Please, place validators for common types (such as uuid, ksuid, cidr, timeout) under https://github.com/vk-cs/terraform-provider-vkcs/tree/master/vkcs/internal/framework/validators.

**Review by paaanic — CHANGES_REQUESTED** [review]

File: vkcs/internal/framework/validators/uuid.go
Comment (cantylv) [reply vkcs/internal/framework/validators/uuid.go:1]: placed

File: vkcs/provider/provider.go (hunk 35 above)
Comment (cantylv) [reply vkcs/provider/provider.go]: done

File: vkcs/kubernetes/resource_cluster_v2_test.go (hunk 34 above)
Comment (cantylv) [reply vkcs/kubernetes/resource_cluster_v2_test.go]: сurrently, service managed-k8s doesn't support cluster scaling without worker nodes, so i've added worker group to the test. API returns an error if the user tries to upgrade the cluster version with no node groups.

File: vkcs/kubernetes/resource_node_group_v2.go (hunk 33 above)
Comment (cantylv) [reply vkcs/kubernetes/resource_node_group_v2.go]: fixed

File: vkcs/internal/clients/config.go [hunk 36]
```diff
@@ -106,6 +107,19 @@ func (c *config) ContainerInfraV1Client(region string) (*gophercloud.ServiceClie
...
 }
 
+func (c *config) ManagedK8SClient(region string) (*gophercloud.ServiceClient, error) {
+	client, err := c.initClient(newManagedK8S, region, "managed-k8s")
+	if err != nil {
+		return client, err
+	}
+
+	client.MoreHeaders = map[string]string{
+		"X-Created-By": "Terraform",
```
Comment (paaanic) [inline vkcs/internal/clients/config.go]: Why is this needed? We already specify User-Agent header, isn't this enough?

**Comment by paaanic** [comment]
Should fix the linting issue.

File: vkcs/internal/clients/config.go (hunk 36 above)
Comment (cantylv) [reply vkcs/internal/clients/config.go]: I want to collect metrics on created clusters through Terraform

**Review by paaanic — APPROVED** [review]
